bazaarvoice/emodb
# Partitioned the fanout queues

## Github Issue #

None

## What Are We Doing Here?

At scale we have found that Emo holds up well to high write traffic but the fanout queues lag quickly under periods of high writes.  Fanout optimizations aside, because this process is run on a single server via leadership election improving fanout speed can only be accomplished by scaling vertically, and even this has bottlenecks which cannot be addressed by vertical scaling alone, such as IO time to Cassandra.

This PR changes the fanout architecture so use mutiple partitions.  As updates are made to Emo or fanned out to remote data centers the events are spread across multiple partitions.  Each partition has a corresponding fanout worker, and which worker executes it is determined through cooperative distributed leadership election.

This PR builds upon https://github.com/bazaarvoice/emodb/pull/154, which itself makes up the first 2 commits of this PR.  To view only the changes relevant to distributed fanout see the final commit.

To maintain a smooth migration this PR includes a "legacy" fanout which will continue to fan out the old master and data center channels.  Once they are fully drained the legacy code can be removed from Emo in a future PR.  In the meantime this will result in nominal overhead as once-per-second the fanout will poll the perpetually empty legacy event queue then go back to sleep.

## How to Test and Verify

There is really no way to "see" that the fanout partitioner is working.  The best way to test is to run Emo through a battery of regression and integration tests and verify that no databus events are lost and that fanout behaves exactly as before.

## Risk

The biggest risk is that an error in fanout could result in lost databus events.  However, since the fanout code itself is unchanged, any likely errors would result from problems with leadership election or an undetected race condition caused by multiple concurrent fanout processes on different hosts.

### Level 

`High`

### Required Testing

`Regression`

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
